kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,308 @@
+    {
+      "id": <analysis-id>,
+      "message": <diagnostic>
+    }
+    ```
+
+The basic types referenced above are:
+
+-   `<analysis-id>`: an integer ID assigned by the server (JSON: number).
+-   `<bytes>`: a base64-encoded byte string (JSON: string).
+-   `<diagnostic>`: a `kythe.proto.common.Diagnostic` message (JSON: object).
+-   `<digest>`: a lower-case hex-coded SHA-256 digest (JSON: string).
+-   `<file-path>`: a file path (JSON: string).
+-   `<protocol>`: a protocol version label (JSON: string).
+-   `<analysis-type>`: an identifier for a specific type or types of analyses.
+-   `<output-encoding>`: an identifier indicating how the output will be encoded, must be one of `json` or `protobuf` (JSON: string).
+
+Where the description refers to a protocol buffer message, the message is
+encoded using the canonical
+[JSON mapping](https://developers.google.com/protocol-buffers/docs/proto3#json)
+for the named message type.
+
+#### analysis-type
+
+Analysis types are described by the following grammar:
+
+```grammatical-framework
+analysis-type         = analysis-type-segment 0*{"/" analysis-type-segment}
+                
+analysis-type-segment = unreserved -- RFC 3987
+```
+ 
+Analysis types whose leading segment is `kythe` are reserved for use by
+Kythe. An example analysis type might be `kythe/index/java`, indicating the
+Kythe indexer for Java.
+
+### Methods
+
+A driver defines the following JSON-RPC methods:
+
+-   `init(InitRequest) → InitReply`
+
+    The `init` method is a protocol handshake, and must be the first method
+    called by the analyzer upon startup. The request specifies the protocol
+    version understood by the analyzer. If the environment understands the
+    requested version, it echoes the version in its reply.
+
+    If any other method is called prior to a successful invocation of `init`, or
+    if the environment doesn't recognize the protocol version requested, the
+    environment will report ProtocolError and close the input channel.
+
+    This document defines the protocol `"kythe1"`.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":1,"method":"init","params":{"protocol":"kythe1", "output-encoding":"json"}}
+        « {"jsonrpc":"2.0","id":1,"result":{"protocol":"kythe1"}}
+
+-   `analyze(AnalyzeRequest) → AnalyzeReply`
+
+    Request a fresh compilation record from the environment, blocking until one
+    is available. The environment assigns a unique integer analysis ID to each
+    request, distinct from any other ID currently assigned to that analyzer.
+    Once an analysis ID has been assigned, it is designated as **pending** until
+    the analyzer sends a `done` notification (see below) for that ID. A driver
+    SHOULD NOT issue the same compilation record again while there is a pending
+    analysis ID for that record.
+    
+    The reply includes the "root" file on which to begin analysis, as well as
+    the output file to which to emit output. The driver (and driver environment)
```

Comment:
Elaborate on the format of the output file and how it depends on the analysis request.

Add an explicit example for Kythe analyses.  Also, at least note what "other" analyzer may do as well.

Reply:
Let me know what you think here.